packages/@react-spectrum/s2/src/TagGroup.tsx

```diff
@@ -86,30 +97,152 @@ const helpTextStyles = style({
+  let showActions = tagState.showCollapseButton || tagState.visibleTagCount < collection.size;
   let formContext = useContext(FormContext);
-  props = useFormProps(props);
-  let {size = 'M'} = props;
   let domRef = useDOMRef(ref);
 
+  let allItems = useMemo(
+    () => Array.from(collection) as Array<Node<T>>,
+    [collection]
+  );
+  let items = useMemo(
+    () => Array.from(collection).slice(0, !isCollapsed ? collection.size : tagState.visibleTagCount) as Array<Node<T>>,
+    [collection, tagState.visibleTagCount, isCollapsed]
+  );
+
+  let updateVisibleTagCount = useEffectEvent(() => {
+    if (maxRows != null && maxRows > 0) {
+      let computeVisibleTagCount = () => {
+        let currContainerRef: HTMLDivElement | null = hiddenTagsRef.current;
+        let currTagsRef: HTMLDivElement | null = hiddenTagsRef.current;
+        let currActionsRef: HTMLDivElement | null = actionsRef.current;
+        if (!currContainerRef || !currTagsRef || collection.size === 0 || currContainerRef.parentElement == null) {
+          return {
+            visibleTagCount: 0,
+            showCollapseButton: false
+          };
+        }
+
+        // Count rows and show tags until we hit the maxRows.
+        // I think this is still a safe assumption, and we don't need to queryAll for role=tag
+        let tags = [...currTagsRef.children];
+        let currY = -Infinity;
+        let rowCount = 0;
+        let index = 0;
+        let tagWidths: number[] = [];
+        for (let tag of tags) {
+          let {width, y} = tag.getBoundingClientRect();
+
+          if (y !== currY) {
+            currY = y;
+            rowCount++;
+          }
+
+          if (rowCount > maxRows) {
+            break;
+          }
+          tagWidths.push(width);
+          index++;
+        }
+
+        // Remove tags until there is space for the collapse button and action button (if present) on the last row.
+        let buttons = currActionsRef ? [...currActionsRef.children] : [];
+        if (buttons.length > 0 && rowCount >= maxRows) {
+          let buttonsWidth = buttons.reduce((acc, curr) => acc += curr.getBoundingClientRect().width, 0);
+          let margins = parseFloat(getComputedStyle(buttons[0]).marginInlineStart);
+          buttonsWidth += margins * 2;
+          let end = direction === 'ltr' ? 'right' : 'left';
+          let containerEnd = currContainerRef.parentElement?.getBoundingClientRect()[end] - margins;
+          let lastTagEnd = tags[index - 1]?.getBoundingClientRect()[end];
+          lastTagEnd += margins;
+          let availableWidth = containerEnd - lastTagEnd;
+
+          while (availableWidth <= buttonsWidth && index > 0) {
+            let tagWidth = tagWidths.pop();
+            if (tagWidth != null) {
+              availableWidth += tagWidth;
+            }
+            index--;
+          }
+        }
+
+        return {
+          visibleTagCount: Math.max(index, 1),
+          showCollapseButton: index < collection.size
```

Comment:
Since the Collapse/Expand button and the groupAction button are in the same group, this logic makes combined with `tagState.visibleTagCount < collection.size` means that the groupAction button will be removed along side the collaspe/expand button if all the tags are visible.

Perhaps the group should always be rendered and `showActions` should instead hide the expand/collapse button from inside the group?

Reply:
pushed a fix